crates/ruff_python_parser/src/lexer.rs

```diff
@@ -508,6 +511,138 @@ impl<'source> Lexer<'source> {
+                    return Err(LexicalError {
+                        error: LexicalErrorType::FStringError(error),
+                        location: self.offset(),
+                    });
+                }
+                '\n' if !context.allow_multiline() => {
+                    self.fstring_stack.pop();
+                    return Err(LexicalError {
+                        error: LexicalErrorType::FStringError(FStringErrorType::UnterminatedString),
+                        location: self.offset(),
+                    });
+                }
+                '\\' => {
+                    self.cursor.bump(); // '\'
+                    if matches!(self.cursor.first(), '{' | '}') {
+                        // Don't consume `{` or `}` as we want them to be consumed as tokens.
+                        break;
+                    } else if !context.is_raw_string {
+                        if self.cursor.eat_char('N') && self.cursor.eat_char('{') {
+                            in_named_unicode = true;
+                            continue;
+                        }
+                    }
+                    // Consume the escaped character.
+                    self.cursor.bump();
+                }
+                quote @ ('\'' | '"') if quote == context.quote_char() => {
+                    match context.quote_size {
+                        StringQuoteSize::Single => break,
+                        StringQuoteSize::Triple => {
+                            let mut remaining = self.cursor.rest().chars().skip(1);
+                            if remaining.next() == Some(quote) && remaining.next() == Some(quote) {
+                                break;
+                            }
+                        }
+                    }
+                    self.cursor.bump();
+                }
+                '{' => {
+                    if self.cursor.second() == '{' {
+                        self.cursor.bump();
+                        self.cursor.bump();
+                    } else {
+                        break;
+                    }
+                }
+                '}' => {
+                    if in_named_unicode {
+                        in_named_unicode = false;
+                        self.cursor.bump();
+                    } else if self.cursor.second() == '}' && !context.is_in_format_spec() {
+                        self.cursor.bump();
+                        self.cursor.bump();
+                    } else {
+                        break;
+                    }
+                }
+                _ => {
+                    self.cursor.bump();
+                }
+            }
+        }
+        let value_end = self.offset();
+
+        if value_start == value_end {
+            return Ok(None);
+        }
+
+        let value = self.source[TextRange::new(value_start, value_end)]
+            .to_string()
+            .replace("{{", "{")
+            .replace("}}", "}");
```

Comment:
Calling replace twice is `O(2N)` on the string length. This may not matter much considering that most strings are small but can add up. It is also unfortunate that we do this even for strings that we know never contain a `{{` or `}}` sequence. There are two ways of how we can avoid this additional pass over the string:

* Track whether the string contained any `{{` or `}}` and only then call replace
* Implement your own `replace` that replaces both sequences in a single loop 
* Create a normalized string, but only initialize it when you see a `{{` or `}}`. Otherwise use the full text range, see https://github.com/astral-sh/ruff/blob/0cea4975fcfe09716c084c0a18d1b4c4cd9b8f05/crates/ruff_python_formatter/src/expression/string.rs#L607-L679

Reply:
Thanks! I've implemented this in [`2de590d` (#6659)](https://github.com/astral-sh/ruff/pull/6659/commits/2de590d2341dfd7f2f5b86458dc1bcba097ab99d)